### PR TITLE
Refactor FXIOS-10096 [WIP] Content blocking: cleanup, refactors + support for custom JSON lists

### PR DESCRIPTION
**Pull request by mattreaganmozilla**
## :scroll: Tickets
[Jira ticket](https://mozilla-hub.atlassian.net/browse/FXIOS-10096)

## :bulb: Description

Clean-up and refactors to our content blocking code. Primarily this is for improvements to readability and performance but it also adds support for custom-loading content lists; right now this feature is entirely for developer convenience and has no effect for end users, but it will allow us to test custom content blocker lists more easily in the app moving forward.

Please note: all of this code is (to some extent) an ongoing WIP as we are currently exploring improvements to our ad-blocking. Additional changes will likely be forthcoming.

## :pencil: Checklist
You have to check all boxes before merging
- [x] Filled in the above information (tickets numbers and description of your work)
- [x] Updated the PR name to follow our [PR naming guidelines](https://github.com/mozilla-mobile/firefox-ios/wiki/Pull-Request-Naming-Guide)
- [ ] Wrote unit tests and/or ensured the tests suite is passing
- [ ] When working on UI, I checked and implemented accessibility (minimum Dynamic Text and VoiceOver)
- [ ] If needed, I updated documentation / comments for complex code and public methods
- [ ] If needed, added a backport comment (example `@Mergifyio backport release/v120`)

### CODE REVIEW

File: firefox-ios/Client/ContentBlocker/ContentBlocker.swift [hunk 1]
```diff
@@ -215,13 +246,17 @@ class ContentBlocker {
...
-                  let source = try? String(contentsOfFile: path, encoding: .utf8)
-            else {
-                self?.logger.log("Error unwrapping the resource contents", level: .warning, category: .webview)
-                assertionFailure("Error unwrapping the resource contents")
-                return
+        let logger = self.logger
+        DispatchQueue.global().async {
+            var source = ""
+            do {
+                let fileTrimmed = file.hasSuffix(".json") ? String(file.dropLast(5)) : file
```
Comment (thatswinnie) [inline firefox-ios/Client/ContentBlocker/ContentBlocker.swift:253]: Can we do something about this magic number? We should at least add a comment I think.

**Review by thatswinnie — COMMENTED** [review]
Would it be possible to add unit tests for the ad blocker?

File: firefox-ios/Client/ContentBlocker/ContentBlocker.swift [hunk 2]
```diff
@@ -188,9 +196,32 @@ class ContentBlocker {
     }
 
     private func add(contentRuleList: WKContentRuleList, toTab tab: ContentBlockerTab) {
+        print("DBG: adding content rule: \(contentRuleList.identifier ?? "<nil>")")
```
[the remark this answers is not in the document]
Comment (mattreaganmozilla) [reply firefox-ios/Client/ContentBlocker/ContentBlocker.swift:199]: This probably should be an actual info-level log but this is all a WIP so for now I've just removed it; thanks for catching!

File: firefox-ios/Client/ContentBlocker/ContentBlocker.swift (hunk 1 above)
Comment (mattreaganmozilla) [reply firefox-ios/Client/ContentBlocker/ContentBlocker.swift:252]: This is hardly a magic number 😄 (there's nothing particularly mysterious about a file suffix string length). That said I'm happy to pull it out and call into `count` instead; I would hope that would be optimized away by the Swift compiler but even if not it won't impact performance. I'll add a comment also.

**Review by nbhasin2 — APPROVED** [review]

**Review by thatswinnie — APPROVED** [review]

**Review by OrlaM — APPROVED** [review]

**Comment by mattreaganmozilla** [comment]
> Would it be possible to add unit tests for the ad blocker?

@thatswinnie Yes that's definitely on the radar, though I think it will likely be revisited after we have a more concrete plan for the ad-blocking updates.

**Pull request merged**